sbin/anti-evil-maid-lib-tpm1

```diff
@@ -0,0 +1,220 @@
+}
+
+revokefreshness() {
+    # invalidates the freshness token of a specified AEM media (by its
+    # label suffix
+    _suff=$1
+    if _slot=$(suffixtoslot "$_suff"); then
+        message "Revoking freshness token for AEM media w/ suffix '$_suff'..."
+        _pw=$(cat "$TPM_FRESHNESS_PASSWORD_FILE")
+        if tpm_nvwrite -i "$TPM_FRESHNESS_INDEX" \
+                -n "$((_slot * 20))" -s 20 \
+                --password="$_pw" -m "0xff"; then
+            message "Done."
+        else
+            message "Failed!"
+        fi
+    else
+        message "AEM device with label suffix '$_suff' not found in DB!"
+    fi
+}
+resetfreshness() {
+    # invalidates ALL freshness tokens
+    message "Invalidating **ALL** freshness tokens..."
+    _pw=$(cat "$TPM_FRESHNESS_PASSWORD_FILE")
+    if tpm_nvwrite -i "$TPM_FRESHNESS_INDEX" \
+            -s "$((TPM_FRESHNESS_SLOTS * 20))" \
+            --password="$_pw" -m "0xff"; then
+        message "Done."
+    else
+        message "Failed!"
+    fi
+}
+
+destroytpmnvram() {
+    # releases the TPM NVRAM area; TPM owner pw as first argument
+    tpm_nvrelease -i "$TPM_FRESHNESS_INDEX" --pwdo="$1"
+}
+
+listbadpcrs() {
+    # prints those standard PCRs configured to be used via $SEAL which haven't
+    # been extended, output is empty there are no such PCRs
+    _pcrs=$(printf %s "$SEAL" | grep -Eo '\b1[3789]\b') || true
+    grep -E "^PCR-(${_pcrs//$'\n'/|}):( 00| FF){20}" "$SYSFS_TPM_DIR"/pcrs
+}
+
+tpmowned() {
+    # checks whether TPM is already owned, signals results with exit code
+    [ "$(cat "$SYSFS_TPM_DIR"/owned)" -ne 0 ]
+}
+
+provisiontpmid() {
+    # stores TPM ID into an NVRAM entry 
+    _tpm_id_index=$(tpm_id -i)
+    _opw=$(cat "$TPM_OWNER_PASSWORD_FILE")
+    # create a write-once NVRAM area
+    tpm_nvdefine -i "$_tpm_id_index" -s 20 -p "WRITEDEFINE|WRITEALL" \
+                 --pwdo="$_opw"
+    # generate a random ID and write it into NVRAM
+    head -c 20 /dev/random | tpm_nvwrite_stdin -i "$_tpm_id_index" -s 20
+    # lock the area to prevent non-owners from changing ID
+    tpm_nvwrite -i "$_tpm_id_index" -s 0
+}
+
+postprovisioning() {
+    # takes care of updating /var/lib/tpms after a successful provisioning by
+    # provisiontpmid
+    _tpmid=$(tpm_id)
+    mkdir -p "/var/lib/tpms/$_tpmid"
+    systemctl stop tcsd
+    mv "$TPMS_DIR"/unknown/* "$TPMS_DIR/$_tpmid/"
+    rm -rf "$TPMS_DIR/unknown"
+    systemctl start tcsd
+}
+
+checksrkpass() {
+    # checks whether contents of $SRK_PASSWORD_CACHE file is a valid SRK
+    # password, signals result with exit code
+    tpm_sealdata -i /dev/null -o /dev/null < "$SRK_PASSWORD_CACHE"
+}
+
+tpmpcrextend() {
+    # extends a PCR with a hash value of a suitable type
+    _pcr=$1
+    _hash=$2
+    tpm_pcr_extend "$_pcr" "$_hash"
+}
+
+tpmsealdata() {
+    # seals source specified by second argument into destination specified by
+    # the third one, non-empty first argument signifies empty SRK password
+    _nosrkpass=()
```

Comment:
similar as in the other place - arrays are bash feature, ensure that all scripts that source this lib have bash shebang (currently some have just `#!/bin/sh`)

Reply:
Except for `sbin/anti-evil-maid-install` the only other script with `/bin/sh` is `etc/grub.d/19_linux_xen_tboot` and it doesn't source `anti-evil-maid-lib`.